doc: Announcing CloudQuery AWS Plugin Event-based Sync (#13618)
cloudquery/cloudquery: website/pages/blog/announcing-cloudquery-aws-event-based-sync-beta.mdx

---
title: Announcing Closed Beta of CloudQuery AWS Plugin with Event-based Sync
tag: announcement
date: 2023/08/31
description:
author: michal
---

import { BlogHeader } from "../../components/BlogHeader"

<BlogHeader/>

> We are excited to announce the preview of an event-based sync for our [AWS Plugin](https://www.cloudquery.io/docs/plugins/sources/aws/overview)!

Reviewer: ```suggestion
We are excited to announce the preview of an event-based sync for our AWS Plugin!
```

Maybe we should not make this a link, if users don't know about the AWS plugin already this post won't really benefit them

## What is it?

> The way CloudQuery has been initially designed was to sync all data on demand to give you the full control over what to sync, when, and how often. This approach makes CloudQuery easy to set up and use.

Reviewer: ```suggestion
CloudQuery was initially designed to sync all data on demand, giving the user full control over what to sync, when, and how often. This approach makes CloudQuery easy to set up and use.
```

> However, regular sync, even if it runs every hour, is sometimes just not enough to get the accurate picture of what is happening in your environment. Cloud environments are ephemeral - they come and go in just a few minutes and it is really hard to track them and get your costs right. An IP address gets spammed by bots trying to find their way in the moment you make it public. User accounts get created with too many permissions and can get misused in a brief moment.

Reviewer: ```suggestion
However, regular sync, even if it runs every hour, is sometimes just not enough to get the accurate picture of what is happening in your environment. Cloud environments are ephemeral - they come and go in just a few minutes and it is really hard to track them and get your costs right. An IP address gets spammed by bots the moment you make it public. User accounts get created with broad permissions and can get misused in a brief moment.
```

This is where our new event-based sync comes to rescue.

<div style={{marginTop:"2em", align: "center"}}><a href="https://cloudquery.typeform.com/to/Dvdn6P2u" class="btn btn-blue">Sign up for early access</a></div>

## How it works

All events are aggregated by AWS CloudTrail. You can configure a Trail to send management events to a Kinesis Data Stream via CloudWatch Logs. By subscribing to a stream of AWS CloudTrail events in the Kinesis Data stream, CloudQuery can then trigger selective syncs to update just the singular resource that had a configuration change.

> ![Configuring CloudQuery AWS Plugin for event-based sync](/images/blog/announcing-aws-event-based-sync/diagram.png)

Reviewer: Nit: I'm wondering if the diagram should include the databases? Ie. not stop at CloudQuery, and show some of the main databases CQ would sync to

Author: good idea, updated.

*Configuring CloudQuery AWS Plugin for event-based sync*

With this setup, you get the fresh data within a few seconds of it becoming available in CloudTrail. 

Here are the services and events supported at the moment:

| Service | Event |
| -- | -- |
| ec2.amazonaws.com |	AttachInternetGateway |
| ec2.amazonaws.com | AuthorizeSecurityGroupEgress |
| ec2.amazonaws.com | AuthorizeSecurityGroupIngress |
| ec2.amazonaws.com | CreateInternetGateway |
| ec2.amazonaws.com | CreateSecurityGroup |
| ec2.amazonaws.com | CreateSubnet |
| ec2.amazonaws.com | CreateTags |
| ec2.amazonaws.com | CreateVpc |
| ec2.amazonaws.com | DeleteTags |
| ec2.amazonaws.com | DetachInternetGateway |
| ec2.amazonaws.com | ModifySubnetAttribute |
| ec2.amazonaws.com | RevokeSecurityGroupEgress |
| ec2.amazonaws.com | RevokeSecurityGroupIngress |
| ec2.amazonaws.com | RunInstances |
| iam.amazonaws.com | CreateGroup |
| iam.amazonaws.com | CreateRole |
| iam.amazonaws.com | CreateUser |
| rds.amazonaws.com | CreateDBCluster |
| rds.amazonaws.com | CreateDBInstance |

## Getting Started

> This feature is available in closed beta at the moment, [sign up for early access](https://cloudquery.typeform.com/to/Dvdn6P2u).

Reviewer: ```suggestion
This feature is available in closed beta, [sign up for early access](https://cloudquery.typeform.com/to/Dvdn6P2u).
```

(Removing repeat use of the phrase `at the moment` from the previous paragraph)

```

2. Copy the ARN of the Kinesis stream. If you used the CloudFormation template you can run the following command:

```bash
aws cloudformation describe-stacks --stack-name <STACK-NAME> --query "Stacks[].Outputs" --region <DESIRED-REGION>
```

3. Define a `config.yml` file like the one below

```yaml
kind: source
spec:  
	name: "aws-event-based"  
	registry: "local"  
	path: <PATH/TO/BINARY>
	tables:
		- aws_ec2_instances
		- aws_ec2_internet_gateways
		- aws_ec2_security_groups
		- aws_ec2_subnets
		- aws_ec2_vpcs
		- aws_ecs_cluster_tasks
		- aws_iam_groups
		- aws_iam_roles
		- aws_iam_users
		- aws_rds_instances
destinations: ["postgresql"]  
	skip_tables:    
		- aws_iam_group_last_accessed_details    
		- aws_iam_role_last_accessed_details    
		- aws_iam_user_last_accessed_details  
	spec:    
		event_based_sync:      
			- account:          
					local_profile: "<ROLE-NAME>"
				kinesis_stream_arn: "<OUTPUT-FROM-CLOUDFORMATION-STACK>"
```

4. Sync the data!

```bash
cloudquery sync config.yml
```

This will start a long lived process that will only stop when there is an error or you stop the process

> ## Deploying in production

Reviewer: I feel like we could move this section to the documentation for the feature, rather than the blog post, WDYT?

Author: I think this is quite important to manage the expectations. It doesn't really hurt, does it?

@bernays what do you think?

Reviewer: I think it's a little too detailed for this post, but I get that it helps to manage expectations. If we keep it we may just want to call out that these are limitations during the beta phase, and we intend to make this easier to manage in the future.
